### PR TITLE
GH-33926: [Python] DataFrame Interchange Protocol for pyarrow.RecordBatch

### LINKED ISSUE #33926: [Python] DataFrame Interchange Protocol for pyarrow.RecordBatch
### Describe the enhancement requested

Add the implementation of the Dataframe Interchange Protocol for `pyarrow.RecordBatch` also. The protocol is already implemented for `pyarrow.Table`, see https://github.com/apache/arrow/pull/14804.

Python dataframe interchange protocol spec: https://data-apis.org/dataframe-protocol/latest/index.html

### Component(s)

Python

### PR DESCRIPTION
**Pull request by AlenkaF**
### Rationale for this change
Add the implementation of the Dataframe Interchange Protocol for `pyarrow.RecordBatch`. The protocol is already implemented for pyarrow.Table, see https://github.com/apache/arrow/pull/14804.

### Are these changes tested?
Yes, tests are added to:

- python/pyarrow/tests/interchange/test_interchange_spec.py
- python/pyarrow/tests/interchange/test_conversion.py
* Closes: #33926

### CODE REVIEW

**Review by jorisvandenbossche — COMMENTED** [review]
Already looking good! Seeing the code to use RecordBatch reminds me we should add some small conveniences to RecordBatch that already exist on Table and can easily be added to RecordBatch as well

File: python/pyarrow/interchange/from_dataframe.py [hunk 1]
```diff
@@ -74,9 +74,9 @@ def from_dataframe(df: DataFrameObject, allow_copy=True) -> pa.Table:
 
     Returns
     -------
-    pa.Table
+    pa.Table or pa.RecordBatch
     """
-    if isinstance(df, pa.Table):
+    if isinstance(df, (pa.Table, pa.RecordBatch)):
```
[the remark this answers is not in the document]
Comment (AlenkaF) [reply python/pyarrow/interchange/from_dataframe.py:79]: Yeah, that should be the case (`from_dataframe` to return `pa.Table`) and it should therefore be the case for `RecordBatch` also. Will change 👍

File: python/pyarrow/tests/interchange/test_conversion.py [hunk 2]
```diff
@@ -274,6 +368,21 @@ def test_roundtrip_pandas_datetime(unit):
     assert expected_protocol.num_chunks() == result_protocol.num_chunks()
     assert expected_protocol.column_names() == result_protocol.column_names()
 
+    batch = table.to_batches()[0]
+    pandas_df = pandas_from_dataframe(batch)
+    result = pi.from_dataframe(pandas_df)
```
[the remark this answers is not in the document]
Comment (AlenkaF) [reply python/pyarrow/tests/interchange/test_conversion.py:373]: Oh, why haven't I thought of that! :)

WIll check again and remove where the repetition is actually not needed and in other cases I will parametrize the tests for table/recordbatch.

File: python/pyarrow/interchange/dataframe.py [hunk 3]
```diff
@@ -154,19 +162,28 @@ def select_columns(self, indices: Sequence[int]) -> _PyArrowDataFrame:
...
-        return _PyArrowDataFrame(
-            self._df.select(list(indices)), self._nan_as_null, self._allow_copy
-        )
+        if isinstance(self._df, pa.RecordBatch):
+            columns = [self._df.column(i) for i in indices]
+            names = [self._df.schema.names[i] for i in indices]
+            return _PyArrowDataFrame(pa.record_batch(columns, names=names))
+        else:
+            return _PyArrowDataFrame(
+                self._df.select(list(indices)), self._nan_as_null, self._allow_copy
```
[the remark this answers is not in the document]
Comment (AlenkaF) [reply python/pyarrow/interchange/dataframe.py:171]: Will do it in a separate PR (issue https://github.com/apache/arrow/issues/34359)

File: python/pyarrow/tests/interchange/test_conversion.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlenkaF) [reply python/pyarrow/tests/interchange/test_conversion.py]: I have removed the repetition from all the tests in _test_conversion.py_ and updated the tests in _test_interchange_spec.py_ to reflect what was suggested here.

**Review by jorisvandenbossche — APPROVED** [review]
Thanks for the updates, looks good to me!

**Pull request merged**